### ISSUE TITLE
Question about importing ONNX model to TIDL bin

### ISSUE THREAD

**Issue by zengpz** [TexasInstruments/edgeai-modelzoo#1]
Hi TI team,

I am currently working on deploying pytorch detection model on TDA4. When I tried to import ONNX model to TIDL .bin, there are a lot of operators are not supported (I also tried some models provided in this model zoo, but none of them worked, which is strange) Here is the error output:

Current ONNX OpSet Version   : 11

 Unsupported slice - axis parameters, in Slice  ONNX operator Gather is not suported now..  By passing
 ONNX operator Gather is not suported now..  By passing
 ONNX operator Gather is not suported now..  By passing
 ONNX operator Unsqueeze is not suported now..  By passing
 ONNX operator Gather is not suported now..  By passing
 ONNX operator Gather is not suported now..  By passing
 ONNX operator Expand is not suported now..  By passing
 ONNX operator Expand is not suported now..  By passing
 ONNX operator Gather is not suported now..  By passing
 ONNX operator Range is not suported now..  By passing
 ONNX operator Unsqueeze is not suported now..  By passing
 ONNX operator ScatterND is not suported now..  By passing
 ONNX operator Gather is not suported now..  By passing
 ONNX operator Constant is not suported now..  By passing
[libprotobuf FATAL /datalocal1/user/kumar/psdkra/protobuf-3.5.1/src/google/protobuf/repeated_field.h:1522] CHECK failed: (index) < (current_size_):
terminate called after throwing an instance of 'google::protobuf::FatalException'
  what():  CHECK failed: (index) < (current_size_):
Aborted (core dumped)

I am using TIDL SDK 7.3, which should be fine according to the tutorial, so could you please tell how can we solve this problem? or does TIDL SDK 8.x support those operators?

Best,
Zeng

**Comment by mathmanu** [comment]
Are you trying to run these models using onnxruntime or using TIDL directly? With TIDL directly, you will only be able to run the supported layers - and you have to specify the output layer accordingly - beyond which there should not be unsupported layers. 

The other option is to run using our onnxruntime which as the capability to offload the supported layers to TIDL and run other layers in ARM. Please see this repository for mode details: https://github.com/TexasInstruments/edgeai-tidl-tools
(But note that the above repository uses binaries from 8.0 release). Our onnxrumtime can also offload the entire detection model graph to TIDL - but you need to provide an additional prototxt file to optimally offload these models to TIDL. Examples are in edgeai-tidl-tools

To assist you with training of embedded friendly PyTorch models, we have to repositories :
https://github.com/TexasInstruments/edgeai-mmdetection
https://github.com/TexasInstruments/edgeai-torchvision (recently updated)

**Comment by zengpz** [comment]
Hi,

Thank you so much for your reply!

Yes, I have tried using ONNX runtime + TIDL to import my .onnx model, and successed! But there are a lot of *net.bin and *io.bin ( I thought there should be only one net.bin and one io.bin ), but I think those are just sub models which are all required during deployment stage?

Another question is about deployment on TDA4. Once we got the .bin model, is it true that we can only use TIOVX to preprocess the images, run the model and then do post process? Since we found it quite hard to write those code using openvx and C according to the vision_app dl demos. 

Thank you again for your help!

**Comment by mathmanu** [comment]
Please see this page for the edgeai starterkit:
https://www.ti.com/edgeai
https://www.ti.com/tool/download/PROCESSOR-SDK-LINUX-SK-TDA4VM

You have easily build realtime applications either in Python or in C++ using this. Its much easier compared to using TIOVX.

**Comment by mathmanu** [comment]
If the numebr of *net.bin files indicate there are multiple subgraphs created. Such a model will not run optimally on device, but will likely to be slow. You can look at the svg visualization files to understand which part is offloaded. Find out the reason why it is not fully offloaded and take corrective action.

For example most of the detection model examples in edgeai-tidl-tools are fully offloaded to TIDL (including the detection post processing layers).

**Comment by zengpz** [comment]
> Please see this page for the edgeai starterkit: https://www.ti.com/edgeai https://www.ti.com/tool/download/PROCESSOR-SDK-LINUX-SK-TDA4VM
> 
> You have easily build realtime applications either in Python or in C++ using this. Its much easier compared to using TIOVX.

Thank you!

Do you mean that we should not follow the demo project provided by vision_app dl demo (which is a TIOVX based C project) to create our application for demo ADAS usage? If so, could you please tell me what kind of role that TIOVX (OpenVX) is playing? Where do we need it?

I am new to this area, Thank you for your help!

**Comment by mathmanu** [comment]
The StarterKit SDK provides a high level gstreamer interface for ease of use. Underneath, it still uses TIOVX. 

There may be some scenarios where you may want to directly use TIOVX instead of gstreamer - for example in a scenario where you don't want to run Linux at all in the SoC but only use RTOS.

But for most users, we now recommend to use the simple and easy to use interface provided by the StarterKit SDK.

**Comment by mathmanu** [comment]
Also see our edgeai software landing page: https://github.com/TexasInstruments/edgeai

**Comment by zengpz** [comment]
Those links are very helpful, I am looking into it, thank you!
